feat(swarm): add `ConnectionDenied::downcast_ref`

nathanielc opened the pull request:
## Description

Prior to this change it was only possible to downcast a `ConnectionDenied` error when you had ownership of it which isn't the case inside `NetworkBehaviour`s. This change allows downcasting by reference.

See https://github.com/libp2p/rust-libp2p/discussions/4018.

## Change checklist

<!-- Please add a Changelog entry in the appropriate crates and bump the crate versions if needed. See <https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases>-->

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

thomaseizinger commented on misc/connection-limits/src/lib.rs:
```diff
@@ -466,6 +466,57 @@ mod tests {
     }
 
+    #[test]
+    fn max_established_incoming_downcast_ref() {
```
I don't think we need this test. We don't actually control any logic here other than forwarding calls so the typesystem should be good enough :)

nathanielc replied:
Should I remove the test or just leave it be?

thomaseizinger reviewed (commented):
Nice!

Add a changelog entry please.

nathanielc committed:
feat(swarm): add ConnectionDenied::downcast_ref

Prior to this change it was possible to only have a reference to a ConnectionDenied error and not be
able to downcast it to check if the connection-limits produced the error.

thomaseizinger reviewed (approved):
Thanks!
